frontend/src/test/java/com/google/tripmeout/frontend/servlet/TripServletTest.java

```diff
@@ -0,0 +1,127 @@
+package com.google.tripmeout.frontend.servlet;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.mockito.Mockito.when;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.tripmeout.frontend.TripModel;
+import com.google.tripmeout.frontend.error.TripAlreadyExistsException;
+import com.google.tripmeout.frontend.error.TripNotFoundException;
+import com.google.tripmeout.frontend.serialization.GsonTripModelTypeAdapter;
+import com.google.tripmeout.frontend.servlet.TripServlet;
+import com.google.tripmeout.frontend.storage.InMemoryTripModelStorage;
+import java.io.IOException;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+public class TripServletTest {
+  private Gson gson;
+  private final static TripModel PARIS = TripModel.builder()
+                                             .setId("CDG")
+                                             .setName("Paris, France")
+                                             .setUserId("12345")
+                                             .setLocationLat(48.3288)
+                                             .setLocationLong(34.0)
+                                             .build();
+
+  @Mock HttpServletRequest request;
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    this.gson = new GsonBuilder()
+                    .registerTypeAdapter(TripModel.class, new GsonTripModelTypeAdapter())
+                    .create();
+  }
+
+  @Test
+  public void returnsErrorCodeForBadIdForDoGet()
+      throws ServletException, IOException, TripNotFoundException, TripAlreadyExistsException {
+    InMemoryTripModelStorage storage = new InMemoryTripModelStorage();
+    storage.addTrip(PARIS);
+    FakeHttpServletResponse response = new FakeHttpServletResponse();
+    TripServlet servlet = new TripServlet(storage, gson);
+
+    when(request.getRequestURI()).thenReturn("/trips/Not_CDG");
+
+    servlet.doGet(request, response);
+    assertThat(response.getStatus()).isEqualTo(404);
```

Comment:
Please use the appropriate constant from `HttpServletResponse` rather than hard coding the code itself. In this case, you'll want `HttpServletResponse.SC_NOT_FOUND`. Same applies anywhere else you're setting a response code or testing it.

Reply:
imo you want the opposite. the SC_NOT_FOUND is stable, but usually hard coding customer experienced behavior in tests is the right thing to do. It forces you to think about how it might break a customer if a constant changes.

Reply:
Eh. I would buy that if there were any possibility it could change, but it's part of a stable API. We're just giving a name to the number. I won't insist on this if you disagree though.

Reply:
Yeah, it's really just preference. Like we both said the response code constants in `HttpServletResponse` are stable, this is really the exceptional case though because of that stability. I'd prefer to keep everything that is part of the user-facing servlet as constants just to keep the everything assuming the "normal" case (not the exceptional case of having a ridiculously stable API)